File: lib/src/internal/OSSAsyncTask.java
package com.alibaba.sdk.android.oss.internal;

import com.alibaba.sdk.android.oss.OSSClientException;
import com.alibaba.sdk.android.oss.OSSServiceException;
import com.alibaba.sdk.android.oss.model.OSSResult;
import com.alibaba.sdk.android.oss.network.ExecutionContext;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

/**
 * Created by zhouzhuo on 11/23/15.
 */
 class OSSAsyncTask<T extends OSSResult> {

     Future<T> future;

     ExecutionContext context;

     volatile bool canceled;

     static OSSAsyncTask wrapRequestTask(Future future, ExecutionContext context) {
        OSSAsyncTask asynTask = OSSAsyncTask();
        asynTask.future = future;
        asynTask.context = context;
        return asynTask;
    }

    /**
     * Cancel the task
     */
     void cancel() {
        canceled = true;
        if (context != null) {
            context.getCancellationHandler().cancel();
        }
    }

    /**
     * Checks if the task is complete
     *
     * @return
     */
     bool isCompleted() {
        return future.isDone();
    }

    /**
     * Waits and gets the result.
     *
     * @return
     * @throws OSSClientException
     * @throws OSSServiceException
     */
     T getResult()  {
        try {
            T result = future.get();
            return result;
        } catch (InterruptedException e) {
            throw OSSClientException(" InterruptedException and message : " + e.getMessage(), e);
        } catch (ExecutionException e) {
            Throwable cause = e.getCause();
            if (cause instanceof OSSClientException) {
                throw (OSSClientException) cause;
            } else if (cause instanceof OSSServiceException) {
                throw (OSSServiceException) cause;
            } else {
                cause.printStackTrace();
                throw OSSClientException("Unexpected exception!" + cause.getMessage());
            }
        }
    }

    /**
     * Waits until the task is finished
     */
     void waitUntilFinished() {
        try {
            future.get();
        } catch (Exception ignore) {
        }
    }

    /**
     * Gets the flag if the task has been canceled.
     */
     bool isCanceled() {
        return canceled;
    }
}

File: lib/src/internal/oss_async_task.dart
import 'package:aliyun_oss_dart_sdk/src/client_exception.dart';
import 'package:aliyun_oss_dart_sdk/src/exception/lib_exception.dart';
import 'package:aliyun_oss_dart_sdk/src/model/oss_result.dart';
import 'package:aliyun_oss_dart_sdk/src/network/lib_network.dart';
import 'package:aliyun_oss_dart_sdk/src/service_exception.dart';

class OSSAsyncTask<T extends OSSResult> {
  Future<T> future;

  /// TODO:使用dio的CancelToken替代
  ExecutionContext? context;

  bool _canceled = false;
  bool _isCompleted = false;

  OSSAsyncTask._(this.future, this.context) {
    future.whenComplete(() {
      _isCompleted = true;
    });
  }

  static OSSAsyncTask<T> wrapRequestTask<T extends OSSResult>(
      Future<T> future, ExecutionContext context) {
    OSSAsyncTask<T> asynTask = OSSAsyncTask._(future, context);
    return asynTask;
  }

  /// Cancel the task
  void cancel() {
    _canceled = true;
      context?.cancellationHandler.cancel();
  }

  /// Checks if the task is complete
  bool isCompleted() => _isCompleted;

  /// Waits and gets the result.
  Future<T> getResult() async {
    try {
      T result = await future;
      return result;
    } on InterruptedException catch (e) {
      throw OSSClientException(
          " InterruptedException and message : $e");
    } catch (e) {
      if (e is OSSClientException || e is OSSServiceException) {
        rethrow;
      } else {
        throw OSSClientException("Unexpected exception! $e");
      }
    }
  }

  /// Waits until the task is finished
  void waitUntilFinished() async {
    try {
      await future;
    } catch (ignore) {
      ///
    }
  }

  /// Gets the flag if the task has been canceled.
  bool isCanceled() => _canceled;
}
